File: LRSMessageToolBar/Classes/InputBar/LRSPlaceholderTextView.m

#import "LRSPlaceholderTextView.h"

@interface LRSPlaceholderTextView()
@property (nonatomic, strong) UILabel *placeHolderLabel;
@end

@implementation LRSPlaceholderTextView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self setup];
        [self addSubview:self.placeHolderLabel];
    }
    return self;
}

#pragma mark - Setters

- (void)setPlaceHolder:(NSString *)placeHolder {
    if ([placeHolder isEqualToString:_placeHolder]) {
        return;
    }
    _placeHolder = placeHolder;
    self.placeHolderLabel.text = _placeHolder;
}

- (void)setPlaceHolderTextColor:(UIColor *)placeHolderTextColor {
    if ([placeHolderTextColor isEqual:_placeHolderTextColor]) {
        return;
    }

    _placeHolderTextColor = placeHolderTextColor;
    self.placeHolderLabel.textColor = _placeHolderTextColor;
}

#pragma mark - Text view overrides

- (void)setText:(NSString *)text {
    [super setText:text];
    self.placeHolderLabel.hidden = !([self.text length] == 0 && self.placeHolder);
}

- (void)setAttributedText:(NSAttributedString *)attributedText {
    [super setAttributedText:attributedText];
    self.placeHolderLabel.hidden = !([self.text length] == 0 && self.placeHolder);
}

- (void)setFont:(UIFont *)font {
    [super setFont:font];
    self.placeHolderLabel.font = font;
}

#pragma mark - Notifications

- (void)didReceiveTextDidChangeNotification:(NSNotification *)notification {
    self.placeHolderLabel.hidden = !([self.text length] == 0 && self.placeHolder);
}

#pragma mark - Life cycle

- (void)setup {
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(didReceiveTextDidChangeNotification:)
                                                 name:UITextViewTextDidChangeNotification
                                               object:self];

    _placeHolderTextColor = [UIColor lightGrayColor];
    _placeHolder = @"";
}

- (void)dealloc {
    _placeHolder = nil;
    _placeHolderTextColor = nil;
    [[NSNotificationCenter defaultCenter] removeObserver:self name:UITextViewTextDidChangeNotification object:self];
}


- (UILabel *)placeHolderLabel {
    if (!_placeHolderLabel) {
        _placeHolderLabel = [[UILabel alloc] initWithFrame:CGRectMake(10, 6, [UIScreen mainScreen].bounds.size.width, 20)];
        _placeHolderLabel.textColor = [UIColor lightGrayColor];
        _placeHolderLabel.font = [UIFont systemFontOfSize:16.0f];
    }
    return _placeHolderLabel;
}
@end

File: LRSMessageToolBar/Classes/InputBar/LRSPlaceholderTextView.swift
 //
//  LRSPlaceholderTextView.swift
//  LRSMessageToolBarComponents
//
//  Created by sama 刘 on 2021/11/16.
//

import UIKit
import Then

public class LRSPlaceholderTextView: UITextView {

    var placeHolder: String? {
        set {
            placeHolderLabel.text = newValue
        }
        get {
            return placeHolderLabel.text
        }
    }
    var placeHolderColor: UIColor? {
        set {
            placeHolderLabel.textColor = newValue
        }
        get {
            return placeHolderLabel.textColor
        }
    }
    lazy var placeHolderLabel: UILabel = UILabel(frame: CGRect(x: 10, y: 0, width: UIScreen.main.width, height: 20)).then{
        $0.textColor = .lightGray
        $0.font = .systemFont(ofSize: 16)
    }

    public override var font: UIFont? {
        willSet {
            placeHolderLabel.font = newValue
        }
    }

    public override var text: String! {
        willSet {
            guard let value = newValue else {
                return
            }
            placeHolderLabel.isHidden = value.count > 0
        }
    }

    override init(frame: CGRect, textContainer: NSTextContainer?) {
        super.init(frame: frame, textContainer: textContainer)
        addSubview(placeHolderLabel)
        NotificationCenter.default.addObserver(forName: UITextView.textDidChangeNotification, object: self, queue: .main) {[weak self] noti in
            if let textCount = self?.text.count, let placeHodlerCount = self?.placeHolder?.count {
                self?.placeHolderLabel.isHidden = !(textCount == 0 && placeHodlerCount != 0)
            }
        }
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

}
